### PR TITLE
DEV: time series changes for 7.4 RC1

### PR DESCRIPTION
**Pull request by dwdougherty**
Ticket: [DOC-3839](https://redislabs.atlassian.net/browse/DOC-3839)

Note to reviewers: There's currently a placeholder (@LIOR NEED) on the time series configuration page . I'll need this release number before these changes can go live.

[DOC-3839]: https://redislabs.atlassian.net/browse/DOC-3839?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: content/commands/ts.add/index.md [hunk 1]
```diff
@@ -83,10 +83,11 @@ stack_path: docs/data-types/timeseries
 summary: Append a sample to a time series
 syntax: "TS.ADD key timestamp value \n  [RETENTION retentionPeriod] \n  [ENCODING\
   \ <COMPRESSED|UNCOMPRESSED>] \n  [CHUNK_SIZE size] \n  [DUPLICATE_POLICY policy] \n  [ON_DUPLICATE policy_ovr] \n\
+  \  [IGNORE ignoreMaxTimediff ignoreMaxValDiff] \n\
   \  [LABELS [label value ...]]\n"
 syntax_fmt: "TS.ADD key timestamp value [RETENTION\_retentionPeriod]\n  [ENCODING\_\
   <COMPRESSED | UNCOMPRESSED>] [CHUNK_SIZE\_size]\n  [DUPLICATE_POLICY\_policy] \n  [ON_DUPLICATE\_<BLOCK | FIRST\
-  \ | LAST | MIN | MAX | SUM>]\n  [LABELS\ [label value ...]]"
+  \ | LAST | MIN | MAX | SUM>]\n  [IGNORE ignoreMaxTimediff ignoreMaxValDiff]\n  [LABELS\ [label value ...]]"
 syntax_str: "timestamp value [RETENTION\_retentionPeriod] [ENCODING\_<COMPRESSED\
```
Comment (LiorKogan) [inline content/commands/ts.add/index.md:91]: I'm not sure what syntax_str is for, but don't we need IGNORE here as well?

Comment (dwdougherty) [reply content/commands/ts.add/index.md:91]: I don't know what it's for either. :) I only update the fields that are necessary. All of that now useless front matter is going away as soon as I get to it anyway.

File: content/commands/ts.add/index.md [hunk 2]
```diff
@@ -177,6 +178,30 @@ This override is effective only for this single command and does not set the tim
 This argument has no effect when a new time series is created by this command.
 </details>
 
+<details open><summary><code>IGNORE ignoreMaxTimediff ignoreMaxValDiff</code></summary> 
+
+is the policy for handling insertion of samples that may be duplicates of existing samples according to the following rules:
+
+  - `ignoreMaxTimeDiff`: the system considers a new insertion a duplicate if the time difference exceeds the maximum allowed. This value defaults to the `IGNORE_MAX_TIME_DIFF` configuration parameter.
```
Comment (LiorKogan) [inline content/commands/ts.add/index.md:185]: It is the other way around. It is considered a duplicate if it DOES NOT exceed both the time difference and the value difference.

Comment (dwdougherty) [reply content/commands/ts.add/index.md:185]: Okay. Updated in next push.

File: content/develop/data-types/timeseries/configuration.md [hunk 3]
```diff
@@ -53,6 +53,8 @@ The following table summarizes which configuration parameters can be set at modu
 | [ENCODING](#encoding) (since RedisTimeSeries v1.6)                         | :white_check_mark: | :white_large_square: |
 | [CHUNK_SIZE_BYTES](#chunk_size_bytes)                                      | :white_check_mark: | :white_large_square: |
 | [OSS_GLOBAL_PASSWORD](#oss_global_password) (since RedisTimeSeries v1.8.4) | :white_check_mark: | :white_large_square: |
+| [IGNORE_MAX_TIME_DIFF](#ignore_max_time_diff) (@LIOR NEED)                 | :white_check_mark: | :white_large_square: |
```
[the remark this answers is not in the document]
Comment (dwdougherty) [reply content/develop/data-types/timeseries/configuration.md:56]: Thanks. Updated in next push.

**Comment by dwdougherty** [comment]
Opening up for greater docs team review.

**Review by cmilesb — APPROVED** [review]
Looks good to me!

**Comment by dwdougherty** [comment]
Thank you, @cmilesb !

File: content/commands/ts.add/index.md [hunk 4]
```diff
@@ -177,6 +178,30 @@ This override is effective only for this single command and does not set the tim
 This argument has no effect when a new time series is created by this command.
 </details>
 
+<details open><summary><code>IGNORE ignoreMaxTimediff ignoreMaxValDiff</code></summary> 
+
+is the policy for handling insertion of samples that may be duplicates of existing samples according to the following rules:
+
+  - `ignoreMaxTimeDiff`: the system considers a new insertion a duplicate if the time difference does not exceed the maximum allowed. This value defaults to the `IGNORE_MAX_TIME_DIFF` configuration parameter.
```
Comment (LiorKogan) [inline content/commands/ts.add/index.md:185]: Please explain these two parameters together. It is considered a duplicate only if both conditions hold: "A new sample is considered duplicate only if ... and ... 

Also, consider not using "the system".

File: content/commands/ts.add/index.md [hunk 5]
```diff
@@ -177,6 +178,30 @@ This override is effective only for this single command and does not set the tim
...
+  - `ignoreMaxValDiff`: the system considers a new insertion a duplicate if the value difference does not exceed the maximum allowed. This value defaults to the `IGNORE_MAX_VAL_DIFF` configuration parameter.
```
Comment (LiorKogan) [inline content/commands/ts.add/index.md:186]: consider changing "value difference" to "absolute value difference" or "absolute value of the difference" between this new sample and the previously reported sample.

File: content/commands/ts.add/index.md [hunk 6]
```diff
@@ -177,6 +178,30 @@ This override is effective only for this single command and does not set the tim
 This argument has no effect when a new time series is created by this command.
 </details>
 
+<details open><summary><code>IGNORE ignoreMaxTimediff ignoreMaxValDiff</code></summary> 
+
+is the policy for handling insertion of samples that may be duplicates of existing samples according to the following rules:
```
Comment (LiorKogan) [inline content/commands/ts.add/index.md:183]: "duplicate of existing samples" may be misleading. We only compare the new sample with the most recent (highest timestamp) sample (see logical condition).
(This means that duplicates would be filtered only if samples are reported in-order).

File: content/commands/ts.add/index.md [hunk 7]
```diff
@@ -177,6 +178,30 @@ This override is effective only for this single command and does not set the tim
...
+
+  - `ignoreMaxTimeDiff`: the system considers a new insertion a duplicate if the time difference does not exceed the maximum allowed. This value defaults to the `IGNORE_MAX_TIME_DIFF` configuration parameter.
+  - `ignoreMaxValDiff`: the system considers a new insertion a duplicate if the value difference does not exceed the maximum allowed. This value defaults to the `IGNORE_MAX_VAL_DIFF` configuration parameter.
+
+  When not specified: set to the global [IGNORE_MAX_TIME_DIFF]({{< baseurl >}}/develop/data-types/timeseries/configuration#ignore_max_time_diff) and [IGNORE_MAX_VAL_DIFF]({{< baseurl >}}/develop/data-types/timeseries/configuration#ignore_max_val_diff), which are, by default, both set to 0.
+
+  These parameters are used when creating a new time series to set the per-key parameters, and are ignored when called with an existing time series (the existing per-key configuration parameters is used).
+
+For each call to `TS.ADD`, if the following condition is met:
```
Comment (LiorKogan) [inline content/commands/ts.add/index.md:192]: Also for each timestamp-value pair reported in `TD.MADD`

File: content/commands/ts.add/index.md [hunk 8]
```diff
@@ -194,7 +219,7 @@ Use it only if you are creating a new time series. It is ignored if you are addi
 
 Returns one of these replies:
 
-- [Integer reply]({{< relref "/develop/reference/protocol-spec#integers" >}}) - the timestamp of the upserted sample
+- [Integer reply]({{< relref "/develop/reference/protocol-spec#integers" >}}) - the timestamp of the upserted sample. For an element that is ignored (see the `IGNORE` parameter above), the value will be `max_timestamp`.
```
Comment (LiorKogan) [inline content/commands/ts.add/index.md:222]: We should explain `max_timestamp` (it is the timestamp of the sample with the largest timestamp in the time series). Also `value_at_max_timestamp` (the value at that `max_timestamp`).

File: content/commands/ts.add/index.md [hunk 9]
```diff
@@ -177,6 +178,30 @@ This override is effective only for this single command and does not set the tim
...
+For each call to `TS.ADD`, if the following condition is met:
+
+```
+if ((series is not a compaction) &&
+    (series' DUPLICATE_POLICY == LAST) &&
+    (timestamp ≥ max_timestamp) &&
+    (timestamp - max_timestamp ≤ ignoreMaxTimeDiff) &&
+    abs(value - value_at_max_timestamp) ≤ ignoreMaxValDiff))
+```
+then ignore this sample.
```
Comment (LiorKogan) [inline content/commands/ts.add/index.md:201]: We should explain `max_timestamp` (it is the timestamp of the sample with the largest timestamp in the time series). Also `value_at_max_timestamp` (the value at that `max_timestamp`).

File: content/develop/data-types/timeseries/configuration.md [hunk 10]
```diff
@@ -246,3 +248,29 @@ Not set
 ```
 $ redis-server --loadmodule ./redistimeseries.so OSS_GLOBAL_PASSWORD password
 ```
+
+### IGNORE_MAX_TIME_DIFF
+
+The system considers a new insertion a duplicate if it occurs within the maximum allowed time difference.
```
Comment (LiorKogan) [inline content/develop/data-types/timeseries/configuration.md:254]: Again, it is a duplicate only when both conditions hold.

File: content/commands/ts.add/index.md [hunk 11]
```diff
@@ -177,6 +178,33 @@ This override is effective only for this single command and does not set the tim
 This argument has no effect when a new time series is created by this command.
 </details>
 
+<details open><summary><code>IGNORE ignoreMaxTimediff ignoreMaxValDiff</code></summary> 
+
+is the policy for handling the insertion of a new sample. Both of the following conditions must be met.
+
+  - The timestamp of the new sample is less than the previous maximum timestamp (`ignoreMaxTimeDiff`).
```
Comment (LiorKogan) [inline content/commands/ts.add/index.md:185]: This is wrong now...
Before, the two conditions were described correctly, though separately, but now - they are wrong.
It is about the difference of the current timestamp from the previous timestamp `(timestamp - max_timestamp ≤ ignoreMaxTimeDiff)`, and about the absolute value difference of the current value from the value and the previous timestamp `abs(value - value_at_max_timestamp) ≤ ignoreMaxValDiff))`.

My latest comment was that we should also mention that this filter is applied only if you add data in-order `(timestamp ≥ max_timestamp)`.

If this is still unclear - let's discuss it.

Comment (dwdougherty) [reply content/commands/ts.add/index.md:185]: Per our Slack conversation, I've patched this up in the next push.

**Review by LiorKogan — APPROVED** [review]

**Pull request merged**